src/org/traccar/reports/ReportUtils.java

```diff
@@ -250,100 +255,79 @@ private static boolean isMoving(ArrayList<Position> positions, int index,
-                            && previousEndParkingIndex < startParkingIndex) {
-                        if (!tripFiltered) {
-                            if (trips) {
-                                result.add(calculateTrip(
-                                        positions, previousEndParkingIndex, startParkingIndex, ignoreOdometer));
-                            } else {
-                                if (result.isEmpty() && previousEndParkingIndex > previousStartParkingIndex) {
-                                    long previousParkingDuration = positions.get(previousEndParkingIndex)
-                                            .getFixTime().getTime() - positions.get(previousStartParkingIndex)
-                                            .getFixTime().getTime();
-                                    if (previousParkingDuration >= tripsConfig.getMinimalParkingDuration()) {
-                                        result.add(calculateStop(positions, previousStartParkingIndex,
-                                                previousEndParkingIndex));
-                                    }
-                                }
-                                result.add(calculateStop(positions, startParkingIndex, isLast ? i : endParkingIndex));
-                            }
-                        }
-                        previousEndParkingIndex = endParkingIndex;
-                        skipped = false;
-                    } else {
-                        skipped = true;
+                if (startTripIndex != -1 && startParkingIndex != -1 && event != null && !deviceState.getMotionState()) {
+                    result.add(calculateTrip(positions, startTripIndex, startParkingIndex, ignoreOdometer));
+                    startTripIndex = -1;
+                }
+            }
+            if (startTripIndex != -1 && startParkingIndex != -1) {
+                result.add(calculateTrip(positions, startTripIndex, startParkingIndex, ignoreOdometer));
+            }
+        }
+        return result;
+    }
+
+    public static Collection<StopReport> detectStops(Collection<Position> positionCollection, TripsConfig tripsConfig,
+            boolean ignoreOdometer, double speedThreshold) {
+        Collection<StopReport> result = new ArrayList<>();
+
+        ArrayList<Position> positions = new ArrayList<>(positionCollection);
+        if (positions != null && !positions.isEmpty()) {
+            MotionEventHandler  motionHandler = new MotionEventHandler(tripsConfig);
+            DeviceState deviceState = new DeviceState();
+            deviceState.setMotionState(isMoving(positions, 0, tripsConfig, speedThreshold));
+            int startTripIndex = -1;
+            int startParkingIndex = deviceState.getMotionState() ? -1 : 0;
+            for (int i = 0; i < positions.size(); i++) {
+                positions.get(i).set(Position.KEY_MOTION, isMoving(positions, i, tripsConfig, speedThreshold));
```

Comment:
The point is that we shouldn't alter positions after they are stored. There must be a better solution.